File: Python/Recocido Simulado.py
import random
import time
import math


# Función de distancia
def distancia(p1, p2):
    return ((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2) ** 0.5


# Función para calcular el largo de un recorrido
def largo_recorrido(ruta, puntos):
    return sum(distancia(puntos[ruta[i]], puntos[ruta[i + 1]]) for i in range(len(ruta) - 1))


# Recocido Simulado
def recocido_simulado(puntos, temperatura_inicial=1000, temperatura_final=1, tasa_enfriamiento=0.995):
    ruta_actual = list(range(len(puntos)))
    random.shuffle(ruta_actual)
    mejor_ruta = ruta_actual[:]
    mejor_largo = largo_recorrido(mejor_ruta, puntos)

    temperatura = temperatura_inicial

    while temperatura > temperatura_final:
        nueva_ruta = ruta_actual[:]
        i, j = random.sample(range(len(puntos)), 2)
        nueva_ruta[i], nueva_ruta[j] = nueva_ruta[j], nueva_ruta[i]

        largo_actual = largo_recorrido(ruta_actual, puntos)
        largo_nueva = largo_recorrido(nueva_ruta, puntos)

        if largo_nueva < largo_actual or random.random() < math.exp((largo_actual - largo_nueva) / temperatura):
            ruta_actual = nueva_ruta

        if largo_nueva < mejor_largo:
            mejor_ruta = nueva_ruta
            mejor_largo = largo_nueva

        temperatura *= tasa_enfriamiento

    return mejor_ruta, mejor_largo


# Ejemplo de uso
if __name__ == "__main__":
    puntos = [(random.randint(0, 100), random.randint(0, 100)) for _ in range(10)]

    start_time = time.time()
    mejor_ruta, mejor_largo = recocido_simulado(puntos)
    end_time = time.time()

    print("Mejor ruta:", mejor_ruta)
    print("Distancia total:", mejor_largo)
    print("Tiempo de ejecución:", end_time - start_time, "segundos")

File: c++/Recocido Simulado.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <algorithm>  // Para std::find, std::iota, std::shuffle
#include <numeric>    // Para iota
#include <random>     // Para random_device, mt19937, uniform_int_distribution
#include <limits>     // Para numeric_limits

using namespace std;

// Función de distancia
double distancia(pair<int, int> p1, pair<int, int> p2) {
    return sqrt(pow(p1.first - p2.first, 2) + pow(p1.second - p2.second, 2));
}

// Función para calcular el largo de un recorrido
double largo_recorrido(const vector<int>& ruta, const vector<pair<int, int>>& puntos) {
    double total = 0;
    for (size_t i = 0; i < ruta.size() - 1; ++i) {
        total += distancia(puntos[ruta[i]], puntos[ruta[i + 1]]);
    }
    return total;
}

// Recocido Simulado
pair<vector<int>, double> recocido_simulado(vector<pair<int, int>>& puntos, double temperatura_inicial = 1000, double temperatura_final = 1, double tasa_enfriamiento = 0.995) {
    vector<int> ruta_actual(puntos.size());
    iota(ruta_actual.begin(), ruta_actual.end(), 0);  // Inicializar la ruta con los índices de los puntos
    random_device rd;
    mt19937 gen(rd());
    shuffle(ruta_actual.begin(), ruta_actual.end(), gen);  // Mezclar la ruta aleatoriamente

    vector<int> mejor_ruta = ruta_actual;
    double mejor_largo = largo_recorrido(mejor_ruta, puntos);
    double temperatura = temperatura_inicial;

    while (temperatura > temperatura_final) {
        vector<int> nueva_ruta = ruta_actual;
        int i = rand() % puntos.size();
        int j = rand() % puntos.size();
        swap(nueva_ruta[i], nueva_ruta[j]);  // Intercambiar dos elementos en la ruta

        double largo_actual = largo_recorrido(ruta_actual, puntos);
        double largo_nueva = largo_recorrido(nueva_ruta, puntos);

        if (largo_nueva < largo_actual || rand() / double(RAND_MAX) < exp((largo_actual - largo_nueva) / temperatura)) {
            ruta_actual = nueva_ruta;
        }

        if (largo_nueva < mejor_largo) {
            mejor_ruta = nueva_ruta;
            mejor_largo = largo_nueva;
        }

        temperatura *= tasa_enfriamiento;
    }

    return make_pair(mejor_ruta, mejor_largo);
}

int main() {
    srand(time(0));

    vector<pair<int, int>> puntos;
    for (int i = 0; i < 10; ++i) {
        puntos.push_back({rand() % 100, rand() % 100});  // Generar puntos aleatorios
    }

    clock_t start_time = clock();

    // Usar 'first' y 'second' para acceder al 'pair'
    pair<vector<int>, double> resultado = recocido_simulado(puntos);
    vector<int> mejor_ruta = resultado.first;
    double mejor_largo = resultado.second;

    clock_t end_time = clock();

    cout << "Mejor ruta:";
    for (int i : mejor_ruta) {
        cout << " " << i;
    }
    cout << "\nDistancia total: " << mejor_largo << "\n";
    cout << "Tiempo de ejecución: " << double(end_time - start_time) / CLOCKS_PER_SEC << " segundos\n";
}
